src/ResourceManager/Common/Commands.ResourceManager.Common/Location/LocationCompleter.cs

```diff
@@ -0,0 +1,200 @@
+        public LocationCompleterAttribute(string[] resourceTypes) : base(CreateScriptBlock(resourceTypes))
+        {
+        }
+
+        public static string[] FindLocations(string[] resourceTypes)
+        {
+            return FindLocations(resourceTypes, ResourceTypeLocationDictionary);
+        }
+
+        /// <summary>
+        /// Finds all valid Locations for 
+        /// </summary>
+        /// <param name="resourceTypes"></param>
+        /// <param name="resourceTypeLocationDictionary"></param>
+        /// <returns></returns>
+        public static string[] FindLocations(string[] resourceTypes, IDictionary<string, ICollection<string>> resourceTypeLocationDictionary)
+        {
+            if (resourceTypeLocationDictionary == null)
+            {
+                resourceTypeLocationDictionary = new Dictionary<string, ICollection<string>>();
+            }
+
+            if (resourceTypes == null)
+            {
+                resourceTypes = new string[0];
+            }
+
+            List<string> validResourceTypes = new List<string>();
+            foreach (string resourceType in resourceTypes)
+            {
+                if (resourceType != null)
+                {
+                    if (resourceTypeLocationDictionary.ContainsKey(resourceType))
+                        validResourceTypes.Add(resourceType);
+#if DEBUG
+                    else throw new Exception(String.Format(Resources.InvalidResourceType, resourceType));
+#endif
+                }
+            }
+
+            string[] distinctLocations;
+            if (validResourceTypes.Count > 0)
+            {
+                distinctLocations = resourceTypeLocationDictionary[validResourceTypes[0]].ToArray();
+                foreach (string resourceType in validResourceTypes)
+                {
+                    distinctLocations = distinctLocations.Intersect(resourceTypeLocationDictionary[resourceType].ToArray(), new LocationEqualityComparer()).ToArray();
+                }
+            }
+#if DEBUG
+            else throw new Exception(Resources.NoValidProviderFound);
+
+            if (distinctLocations.Length < 1) throw new Exception(Resources.NoValidLocationsFound);
+#endif
+            for (int i = 0; i < distinctLocations.Length; i++)
+            {
+                distinctLocations[i] = "\"" + distinctLocations[i] + "\"";
+            }
+            return distinctLocations;
+        }
+
+        /// <summary>
+        /// Create a dictionary mapping the ResourceTypes of all providers given to the valid locations available to that ResourceType
+        /// </summary>
+        /// <param name="allProviders"></param>
+        /// <returns></returns>
+        public static IDictionary<string, ICollection<string>> CreateLocationDictionary(List<Provider> allProviders)
+        {
+            IDictionary<string, ICollection<string>> resourceTypeLocationDictionary = new Dictionary<string, ICollection<string>>(StringComparer.OrdinalIgnoreCase);
```

Comment:
Use ConcurrentDictionary, sicne this could be used by multiple threads.

Reply:
Also, please be explicit abotu the key equality comprator in the constructor

Reply:
Is StringComparer.OrdinalIgnoreCase not explicit?